include/drivers/i2c.h

```diff
@@ -129,6 +129,27 @@ struct i2c_msg {
  */
 struct i2c_slave_config;
 
+/**
+ * @brief Callback function to request emulation of an I2C transfer
+ *
+ * You can either use this callback or all the others below it. This
+ * one is more convenient for native_posix testing since it provides a
+ * higher-level API.
+ *
+ * Check your I2C driver to see which one it implements. If your
+ * simulator needs to be used by multiple I2C drivers, then you may
+ * need to implement both!
+ *
```

Comment:
It does not require existing code to be modified. This is adding a new, simpler call to do a similar this. The STM32 supports 5 separate callbacks which have the same effect but better match how that driver works.

So we could modify STM32 to do things this way, but I'm not sure if there is any benefit. Also other drivers may split things out like STM32?

Reply:
Are you referring to the STM32 HAL, or something in Zephyr?  And to an STM32 I2C simulator, or an STM32 I2C driver?  It looks like this could be used in a driver as an alternative to implementing the finer-grained ones.

AFAICT right now STM32 is the only target that supports the "slave" API, though I think some additional ones might be submitted.  I don't see anything like this in any I2C "slave" driver.

I can say that the original Zephyr I2C model of describing a multi-part transaction with a scatter/gather iovec-like API has caused significant pain, because not all I2C peripherals can support fragmenting bus transactions that way, and the API was not defined with precision so there are cross-driver differences in behavior and flag interpretation.  So the thought of adding another instance of it concerns me.

Though I concede that if presence/absence of an alternative implementation isn't visible to the user it wouldn't be an API change.  But then I'm not sure how anybody would need to implement multiple ones in a simulator.

Maybe I need a better understanding of exactly what you're aiming to do.  I thought it was just to run the existing I2C API tests on native POSIX in addition to real hardware.

Reply:
Yes that's right. But as I mentioned the current API for the slave is too fine-grained and seems to be tied in with STM32. I'd like a simple function like the one I added.

Reply:
> Maybe I need a better understanding of exactly what you're aiming to do

My goal is to have a way to write simulators for all the I2C peripherals, not just EEPROM. So for example we could have simulator (or fakes) for battery, accelerometer, lid sensor, temperature sensor, etc.

Reply:
> My goal is to have a way to write simulators for all the I2C peripherals, not just EEPROM. So for example we could have simulator (or fakes) for battery, accelerometer, lid sensor, temperature sensor, etc.

OK.  So say we have a simulator for a Bosche BME280.  That simulator would presumably be something that hooks into an I2C-like controller simulator that recognizes specific device addresses and passes bus transactions to the corresponding device simulator to fill in response data.  So we'd be testing the real `drivers/sensor/bme280` but on native-posix with a faked I2C controller that's the bus parent of the faked sensor.  I assume `zephyr,i2c-sim` is intended to be that faked controller.

Is that use case in line with what you had in mind?

Maybe the core problem is that this test is intended to verify the I2C "slave" API, which is only implemented on one platform (STM32), and this test relies on a simulated device that in reality wouldn't provide an I2C "slave" interface.  A real EEPROM would be accessed through an I2C controller acting in "master" mode.  The test here checks the I2C "slave" mode implementation on STM32 only because there's a real STM32 I2C controller configured to that mode that uses a hack to inject data in response to driver-specific calls.

I don't think this is a good candidate for demonstrating the value of an I2C simulator.  Something like a BME280 would be better: if `samples/sensor/bme280` would run on native_posix with a BME280 simulator that could be somewhat useful, though I'd still go to real hardware.

Or, sticking with EEPROM, something that pretends to be an AT24C32 and passes the tests for `drivers/eeprom`.  Still a simulated controller on a simulated device, but I guess the production code for the eeprom driver would be tested (within the accuracy limits of the simulation of the device wrt timing and other edge conditions).